machine-runner/src/test-utils/mock-runner.ts

```diff
@@ -0,0 +1,197 @@
+            control.reject()
+          } else {
+            control.resolve()
+          }
+          // eslint-disable-next-line @typescript-eslint/no-empty-function
+          return promise.catch(() => {})
+        }),
+      )
+      data.delayedCommands = []
+    }
+
+    return {
+      make,
+      toggle,
+    }
+  }
+}
+
+export namespace Subscription {
+  export const make = <RegisteredEventsFactoriesTuple extends MachineEvent.Factory.Any[]>() => {
+    type Subscribe = SubscribeFn<RegisteredEventsFactoriesTuple>
+    type Cb = Parameters<Subscribe>[0]
+    const cancel = () => {
+      if (data.cb === null) throw new Error('not subscribed')
+      data.cb = null
+    }
+    const subscribe: Subscribe = (onEvent) => {
+      if (data.cb !== null) throw new Error('already subscribed')
+      data.cb = onEvent
+      return data.cancel
+    }
+
+    const data = {
+      cb: null as null | Cb,
+      cancel,
+      subscribe,
+    }
+    return data
+  }
+}
+
+export const createMockMachineRunner = <
+  SwarmProtocolName extends string,
+  MachineName extends string,
+  RegisteredEventsFactoriesTuple extends MachineEvent.Factory.Any[],
+  Payload,
+>(
+  factory: StateFactory<
+    SwarmProtocolName,
+    MachineName,
+    RegisteredEventsFactoriesTuple,
+    string,
+    Payload,
+    object
+  >,
+  payload: Payload,
+) => {
+  const delayer = PromiseDelay.make()
+  const sub = Subscription.make<RegisteredEventsFactoriesTuple>()
+  const persisted: MachineEvent.Any[] = []
+
+  const feed = (
+    ev: MachineEvent.Factory.ReduceToEvent<RegisteredEventsFactoriesTuple>[],
+    props?: { caughtUp: true },
+  ) => {
+    const caughtUp = props === undefined || props.caughtUp === true
+    if (sub.cb === null) throw new Error('not subscribed')
+    return sub.cb({
+      type: MsgType.events,
+      caughtUp,
+      events: ev.map((payload) => ({
+        meta: mockMeta(),
+        payload,
+      })),
+    })
+  }
+
+  const timeTravel = () => {
+    if (sub.cb === null) throw new Error('not subscribed')
+    const cb = sub.cb
+    cb({ type: MsgType.timetravel, trigger: { lamport: 0, offset: 0, stream: 'stream' } })
+    if (sub.cb === null) throw new Error('did not resubscribe')
+  }
+
+  const assertAs = <
+    N extends string,
+    P,
+    C extends CommandDefinerMap<any, any, MachineEvent.Any[]>,
+    Then extends (state: State<N, P, C>) => any,
+  >(
+    factory: StateFactory<SwarmProtocolName, MachineName, RegisteredEventsFactoriesTuple, N, P, C>,
+    then?: Then,
+  ) => {
+    const opaque = machine.get()
+    if (!opaque) {
+      throw new Error(`MachineRunnerTestError: opaque not retrievable yet`)
+    }
+    const snapshot = opaque.as(factory)
+    if (!snapshot) {
+      throw new Error(
+        `MachineRunnerTestError: expected type ${factory.mechanism.name} found ${opaque.type}`,
+      )
+    }
+    if (then) {
+      return then(snapshot)
+    }
+  }
+  // Below: public
+
+  const delay = {
```

Comment:
This facility is not immediately obvious — could you add some docs to show what it can do?

Reply:
Sure. Will do

Reply:
This tool is originally used to emulate a delayed response from Actyx in our internal test.
On second thought, our users should not need this until they need integration tests, right?

Reply:
Yes, we should remove it from the API for now and wait until we have a clear picture on how to expose such functionality.

Reply:
Ok. Pushing the change now https://github.com/Actyx/machines/pull/30/commits/f762a773b60d6edf30dbd7a1eb42e1f50a0e6b43